### PR TITLE
mat: Add DiagView methods to types

### LINKED ISSUE #213: mat: get diagonal of a matrix
as per discussion with @btracey on slack, it'd very handy to be able to extract the diagonal of a matrix for e. g. extracting the variance from a varcov matrix (cf. `numpy.diag` when operating on a matrix)

possibly as a view directly on the matrix:

```go
func (m *Dense) DiagView() SymBandDense {}
```

or

```
func (m *Dense) DiagView() Dense {}
```

or

```go
func (s *SymBandDense) GetDiag(a Matrix)
```

### PR DESCRIPTION
**Pull request by btracey**
Fixes #213

Please take a look.

### CODE REVIEW

**Comment by btracey** [comment]
This intentionally does not fix the `TODO` in `DiagFrom`. I think there's a much cleaner implementation now that `DiagView` is in.

File: mat/dense_test.go [hunk 1]
```diff
@@ -356,6 +356,26 @@ func TestRowColView(t *testing.T) {
 	}
 }
 
+func TestDiagView(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (btracey) [reply mat/dense_test.go]: Almost none the `Dense` tests have `Dense` in the names.

File: mat/symmetric_test.go [hunk 2]
```diff
@@ -129,6 +129,16 @@ func TestSymAtSet(t *testing.T) {
 	}
 }
 
+func TestSymDiagView(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (btracey) [reply mat/symmetric_test.go]: The Sym tests are all `TestSymXxx`

File: mat/triangular.go [hunk 3]
```diff
@@ -296,6 +296,18 @@ func (t *TriDense) isolatedWorkspace(a Triangular) (w *TriDense, restore func())
 	}
 }
 
+// DiagView returns the diagonal as a matrix backed by the original data.
+func (t *TriDense) DiagView() Diagonal {
+	n := t.mat.N
```
[the remark this answers is not in the document]
Comment (btracey) [reply mat/triangular.go]: Done.

File: mat/triangular_test.go [hunk 4]
```diff
@@ -139,6 +139,19 @@ func TestTriAtSet(t *testing.T) {
 	}
 }
 
+func TestTriDiagView(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (btracey) [reply mat/triangular_test.go]: Almost all of the Tri tests do not have Dense on them

File: mat/triband.go [hunk 5]
```diff
@@ -305,3 +305,24 @@ func (t *TriBandDense) TTriBand() TriBanded {
 func (t *TriBandDense) RawTriBand() blas64.TriangularBand {
 	return t.mat
 }
+
+// DiagView returns the diagonal as a matrix backed by the original data.
+func (t *TriBandDense) DiagView() Diagonal {
+	n := t.mat.N
+	if t.isUpper() {
```
[the remark this answers is not in the document]
Comment (btracey) [reply mat/triband.go]: Nice. Done.

File: mat/triband_test.go [hunk 6]
```diff
@@ -332,3 +332,30 @@ func TestTriBandAtSetUpper(t *testing.T) {
 		}
 	}
 }
+
+func TestTriBandDiagView(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (btracey) [reply mat/triband_test.go]: Same as others.

File: mat/dense_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kortschak) [reply mat/dense_test.go]: This is something I'd like to fix.

File: mat/symmetric_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kortschak) [reply mat/symmetric_test.go]: Same here and the rest.

[the remark this answers is not in the document]
Comment (btracey) [reply mat/symmetric_test.go]: We can do that in a different PR.

File: mat/diagonal.go [hunk 7]
```diff
@@ -170,6 +170,11 @@ func (d *DiagDense) Reset() {
 	d.mat.Data = d.mat.Data[:0]
 }
 
+// DiagView returns the diagonal as a matrix backed by the original data.
+func (d *DiagDense) DiagView() Diagonal {
+	return d
+}
+
 // DiagFrom copies the diagonal of m into the receiver. The receiver must
 // be min(r, c) long or zero. Otherwise DiagOf will panic.
```
Comment (kortschak) [inline mat/diagonal.go]: Can you fix this thing I missed? s/DiagOf/DiagFrom/

Comment (btracey) [reply mat/diagonal.go]: Ah, shoot, Done.

Comment (kortschak) [reply mat/diagonal.go]: Was my fault.

**Comment by btracey** [comment]
Weird. It does seem to have (I have the change locally at git claims it's synced)

**Comment by btracey** [comment]
There we go.

**Pull request merged**